prov/util/src/util_atomic.c

```diff
@@ -136,22 +185,104 @@ size_t ofi_datatype_size(enum fi_datatype datatype)
 			op(type, d[i], s[i]);				\
 	}
 
+#ifdef HAVE_BUILTIN_MM_ATOMICS
+
+#define OFI_DEF_WRITEEXT_NAME(op, type) ofi_write_## op ##_## type,
+#define OFI_DEF_WRITEEXT_FUNC(op, type)					\
+	static void ofi_write_## op ##_## type				\
+		(void *dst, const void *src, size_t cnt)		\
+	{								\
+		type target;						\
+		type val;						\
+		size_t i;						\
+		type *d = (dst);					\
+		const type *s = (src);					\
+		int success;						\
+									\
+		for (i = 0; i < cnt; i++) {				\
+			success = 0;					\
+			do {						\
+				target = d[i];				\
+				val = op(type, d[i], s[i]);		\
+				success = __atomic_compare_exchange(	\
+						&d[i],&target,&val,0,	\
+						__ATOMIC_SEQ_CST,	\
+						__ATOMIC_SEQ_CST);	\
+			} while (!success);				\
+		}							\
+	}
+
+#define OFI_DEF_WRITEEXT_CMP_NAME(op, type) ofi_write_## op ##_## type,
+#define OFI_DEF_WRITEEXT_CMP_FUNC(op, type)				\
+	static void ofi_write_## op ##_## type				\
+		(void *dst, const void *src, size_t cnt)		\
+	{								\
+		type target;						\
+		size_t i;						\
+		type *d = (dst);					\
+		const type *s = (src);					\
+		int success;						\
+									\
+		for (i = 0; i < cnt; i++) {				\
+			do {						\
+				success = 1;				\
```

Comment:
In this case since the code does a logical or arithmetic comparison with the destinations current value, if the comparison if false, we are done and will not attempt the __atomic_compare_exchange(). This is true for all of the xxxEXT_CMP() variants.